src/main/java/frc/robot/subsystems/DrivetrainSubsystem.java

```diff
@@ -226,19 +240,32 @@ public void periodic() {
                 } else {
                         updateRotationTarget = true;
                 }
+
+                if(Robot.isSimulation()) {
+                        if(!DriverStationSim.getAutonomous()) {
+                                states = kinematics.toSwerveModuleStates(chassisSpeeds);
+                        }
+                        positionMeters[0] += states[0].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[1] += states[1].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[2] += states[2].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[3] += states[3].speedMetersPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        simRotation += chassisSpeeds.omegaRadiansPerSecond * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        
+                }else {
+                        positionMeters[0] += frontLeftModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime); // (m / s) * delta t = m
+                        positionMeters[1] += frontRightModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[2] += backLeftModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                        positionMeters[3] += backRightModule.getDriveVelocity() * (Timer.getFPGATimestamp() - lastUpdateTime);
+                }
                 
-                positionMeters[0] += frontLeftModule.getDriveVelocity() * Timer.getFPGATimestamp() - lastUpdateTime; // (m / s) * delta t = m
-                positionMeters[1] += frontRightModule.getDriveVelocity() * Timer.getFPGATimestamp() - lastUpdateTime;
-                positionMeters[2] += backLeftModule.getDriveVelocity() * Timer.getFPGATimestamp() - lastUpdateTime;
-                positionMeters[3] += backRightModule.getDriveVelocity() * Timer.getFPGATimestamp() - lastUpdateTime;
                 lastUpdateTime = Timer.getFPGATimestamp();
 
                 robotPose = odometry.update(
                         getGyroscopeRotation(),
                         getModulePositions()
                 );
 
-                if(chassisSpeeds.omegaRadiansPerSecond + chassisSpeeds.vxMetersPerSecond + chassisSpeeds.vxMetersPerSecond == 0) {
+                if(chassisSpeeds.omegaRadiansPerSecond + chassisSpeeds.vxMetersPerSecond + chassisSpeeds.vxMetersPerSecond != 0 && !DriverStation.isAutonomous()) {
                         states = kinematics.toSwerveModuleStates(chassisSpeeds);
```

Comment:
I would like to limit our conditionals for checking if we are in autonomous and feel this can be solved by moving this line of code to the ```drive(ChassisSpeeds )``` method like we discussed. This method is only used in teleop and therefore removes the need for the conditional.

Reply:
I thought about this a little more and I am not sure my suggestion will work, but I think we need to talk through how to generalize our rotation PID handling a bit.

My high level idea is the following:

* ensure that external to the periodic method we have a known desired set of ```chassisSpeeds```. This would mean in autonomous we would need to call kinematics.toChassisSpeeds in setModuleStates.
* move the handling of setting the rotation target based on the gyro for drift correction to the DefaultDriveCommand as this command will only be used in teleop and helps us not interfere with any of the chassis speeds that might have been set in auto
* At the top of the periodic method, we should check only if the rotationTarget is active (because it was set externally) and if so modify the desired chassisSpeed.omegaRadiansPerSecond , then convert the chassisSpeeds toSwerveModuleStates. Later we can easily extend this to have PIDcontrollers for X and Y translation (could be useful for moving automatically between bot poses).